hpx/parallel/segmented_algorithms/transform.hpp

```diff
@@ -0,0 +1,876 @@
+                    }
+                }
+
+                // handle the beginning of the last partition
+                beg = traits1::begin(sit);
+                end = traits1::local(last);
+                if (beg != end)
+                {
+                    segments.push_back(
+                        dispatch_async(traits1::get_id(sit),
+                            algo, policy, forced_seq(),
+                            beg, end, ldest, f, proj)
+                    );
+                }
+            }
+
+            return result::get(
+                dataflow(
+                    [=](std::vector<future<std::pair<local_iterator_type1,
+                    local_iterator_type2> > > && r) -> std::pair<SegIter, OutIter>
+                    {
+                        // handle any remote exceptions, will throw on error
+                        std::list<std::exception_ptr> errors;
+                        parallel::util::detail::handle_remote_exceptions<
+                            ExPolicy
+                        >::call(r, errors);
+                        auto ft = r.back().get();
+                        auto olast = traits1::compose(send, std::get<0>(ft));
+                        auto odest = traits2::compose(sdest, std::get<1>(ft));
+                        return std::make_pair(olast, odest);
+                    },
+                    std::move(segments)));
+        }
+
+        ///////////////////////////////////////////////////////////////////////
+        // segmented implementation
+        template <typename ExPolicy, typename SegIter, typename OutIter, typename F,
+            typename Proj>
+        typename util::detail::algorithm_result<ExPolicy,
+            hpx::util::tagged_pair<tag::in(SegIter), tag::out(OutIter)>
+        >::type
+        transform_(ExPolicy&& policy, SegIter first, SegIter last, OutIter dest,
+            F && f, Proj && proj, std::true_type)
+        {
+            typedef parallel::execution::is_sequenced_execution_policy<
+                    ExPolicy
+                > is_seq;
+            typedef util::detail::algorithm_result<ExPolicy,
+                hpx::util::tagged_pair<tag::in(SegIter), tag::out(OutIter)>
+                > result;
+            if (first == last)
+            {
+                return result::get(std::make_pair(std::move(last),
+                    std::move(dest)));
+            }
+
+            typedef hpx::traits::segmented_iterator_traits<SegIter>
+                iterator_traits1;
+            typedef hpx::traits::segmented_iterator_traits<OutIter>
+                iterator_traits2;
+
+            return hpx::util::make_tagged_pair<tag::in, tag::out>(
+                segmented_transform(transform<std::pair<
+                typename iterator_traits1::local_iterator,
+                typename iterator_traits2::local_iterator> >(),
+                std::forward<ExPolicy>(policy), first, last,
+                dest, std::forward<F>(f), proj, is_seq()));
```

Comment:
Please add proper indentation to make this snippet more readable. Also, `proj` should be perfectly forwarded:

            return hpx::util::make_tagged_pair<tag::in, tag::out>( 
                 segmented_transform(
                     transform<std::pair<
                         typename iterator_traits1::local_iterator, 
                         typename iterator_traits2::local_iterator
                     > >(), 
                     std::forward<ExPolicy>(policy), first, last, 
                     dest, std::forward<F>(f), std::forward<Proj>(proj), 
                     is_seq()));